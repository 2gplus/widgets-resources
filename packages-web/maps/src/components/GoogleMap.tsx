import { createElement, ReactElement, useEffect, useRef } from "react";
import classNames from "classnames";

import googleApiWrapper from "./GoogleApi";
import { getDimensions, getGoogleMapsStyles } from "../utils";
import { Marker, SharedProps } from "../../typings";

export interface GoogleMapsProps extends SharedProps {
    scriptsLoaded?: boolean;
    mapStyles?: string;

    optionStreetView: boolean;
    mapTypeControl: boolean;
    fullScreenControl: boolean;
    rotateControl: boolean;
}

export const GoogleMap = (props: GoogleMapsProps): ReactElement => {
    const map = useRef<google.maps.Map>();
    const googleMapsRef = useRef<HTMLDivElement>(null);
    const defaultCenterLocation: google.maps.LatLngLiteral = { lat: 51.906688, lng: 4.48837 };
    let bounds!: google.maps.LatLngBounds;

    const markers = useRef<google.maps.Marker[]>([]); // Used to manage and remove markers from the map

    useEffect(() => {
        if (props.scriptsLoaded) {
            createMap();
        }
    }, [props.scriptsLoaded]);

    useEffect(() => {
        addMarkers();
    }, [props.locations, props.currentLocation]);

    const createMap = (): void => {
        const mapOptions: google.maps.MapOptions = {
            zoom: props.zoomLevel,
            zoomControl: props.optionZoomControl,
            scrollwheel: props.optionScroll,
            draggable: props.optionDrag,
            streetViewControl: props.optionStreetView,
            mapTypeControl: props.mapTypeControl,
            fullscreenControl: props.fullScreenControl,
            rotateControl: props.rotateControl,
            minZoom: 1,
            maxZoom: 20,
            styles: getGoogleMapsStyles(props.mapStyles)
        };
        if (googleMapsRef.current && !map.current) {
            map.current = new google.maps.Map(googleMapsRef.current, mapOptions);
            bounds = new google.maps.LatLngBounds().extend(defaultCenterLocation);
        } else if (map.current) {
            map.current.setOptions(mapOptions);
        }
        addMarkers();
    };

    const addMarkers = (): void => {
        markers.current.forEach(marker => marker.setMap(null));
        markers.current = [];
        if (props.locations && props.locations.length > 0) {
            bounds = new google.maps.LatLngBounds();
            markers.current = props.locations.reduce<google.maps.Marker[]>((markerArray, currentLocation) => {
                const marker = addMarker(currentLocation);
                if (marker) {
                    markerArray.push(marker);
                }
                return markerArray;
            }, []);
        }
        addCurrentLocation();
        updateCamera();
    };

    const addCurrentLocation = (): void => {
        if (map.current && props.currentLocation) {
            markers.current = props.locations
                .concat(props.currentLocation ? [props.currentLocation] : [])
                .map(addMarker)
                .filter(m => m);
        }
    };

    const addMarker = (marker: Marker): google.maps.Marker => {
        if (!bounds) {
            bounds = bounds = new google.maps.LatLngBounds();
        }
        bounds.extend({
            lat: marker.latitude,
            lng: marker.longitude
        });
        const mapMarker = new google.maps.Marker({
            position: {
                lat: marker.latitude,
                lng: marker.longitude
            },
            icon: marker.url
                ? {
                      url: marker.url,
                      scaledSize: new google.maps.Size(32, 32)
                  }
                : "",
            title: marker.title
        });
        if (marker.title) {
            const infoContent = document.createElement("span");
            infoContent.innerHTML = marker.title || "";
            if (marker.onClick) {
                infoContent.style.cursor = "pointer";
                infoContent.onclick = marker.onClick;
            }
            const infoWindow = new google.maps.InfoWindow({
                content: infoContent
            });
            mapMarker.addListener("click", () => {
                infoWindow.open(map.current, mapMarker);
            });
        } else {
            if (marker.onClick) {
                mapMarker.addListener("click", () => {
                    marker.onClick!();
                });
            } else {
                mapMarker.setClickable(false);
            }
        }
        mapMarker.setMap(map.current!);
        return mapMarker;
    };

    const updateCamera = (): void => {
        const { zoomLevel, autoZoom } = props;
        if (bounds && map.current) {
            try {
                if (!autoZoom) {
                    map.current.setCenter(getCenter());
                    map.current.setZoom(zoomLevel);
                } else {
                    map.current.fitBounds(bounds);
                }
            } catch (error) {
                console.error(`Invalid map bounds ${error.message}`);
            }
        }
    };

    const getCenter = (): google.maps.LatLngLiteral => ({
        lat: markers.current?.[0]?.getPosition()?.lat() ?? defaultCenterLocation.lat,
        lng: markers.current?.[0]?.getPosition()?.lng() ?? defaultCenterLocation.lng
    });

    return (
        <div className={classNames("widget-maps", props.className)} style={{ ...props.style, ...getDimensions(props) }}>
            <div className="widget-google-maps-wrapper">
                <div className="widget-google-maps" ref={googleMapsRef} />
            </div>
        </div>
    );
};

export default googleApiWrapper("https://maps.googleapis.com/maps/api/js?key=")(GoogleMap);
