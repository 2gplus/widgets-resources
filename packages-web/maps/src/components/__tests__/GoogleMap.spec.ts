import { shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";
import { GoogleMap, GoogleMapsProps } from "../GoogleMap";
import { initialize } from "@googlemaps/jest-mocks";

describe("Google maps", () => {
    const defaultProps: GoogleMapsProps = {
        autoZoom: true,
        height: 75,
        heightUnit: "pixels",
        optionDrag: true,
        optionScroll: true,
        optionZoomControl: true,
        zoomLevel: 10,
        width: 50,
        widthUnit: "percentage",
        mapStyles: "",
        style: {},
        currentLocation: undefined,
        optionStreetView: false,
        locations: [],
        className: "",
        fullScreenControl: false,
        mapsToken: "",
        mapTypeControl: false,
        rotateControl: false,
        showCurrentLocation: false
    };

    beforeEach(() => {
        initialize();
    });

    const renderGoogleMap = (props: GoogleMapsProps): ShallowWrapper<GoogleMapsProps, any> =>
        shallow(createElement(GoogleMap, props));

    it("renders a map with right structure", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "pixels"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with pixels renders structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "pixels",
            widthUnit: "pixels"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with percentage of width and height units renders the structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "percentage"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with percentage of parent units renders the structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfParent",
            widthUnit: "percentage"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with markers", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            locations: [
                {
                    title: "Mendix HQ",
                    latitude: 51.906688,
                    longitude: 4.48837,
                    url: "image:url"
                },
                {
                    title: "Gementee Rotterdam",
                    latitude: 51.922823,
                    longitude: 4.479632,
                    url: "image:url"
                }
            ]
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("renders a map with current location", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            showCurrentLocation: true,
            currentLocation: {
                latitude: 51.906688,
                longitude: 4.48837,
                url: "image:url"
            }
        });

        expect(googleMaps).toMatchSnapshot();
    });

    afterAll(() => {
        window.google = undefined;
    });
});
