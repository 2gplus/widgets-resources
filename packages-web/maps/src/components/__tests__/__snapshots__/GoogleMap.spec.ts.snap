// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google maps renders a map with percentage of parent units renders the structure correctly 1`] = `
<LoadScript
  googleMapsApiKey=""
  id="_com.mendix.widget.custom.Maps.Maps"
  onLoad={[Function]}
  version="weekly"
>
  <div
    className="widget-maps"
    style={
      Object {
        "height": "75%",
        "width": "50%",
      }
    }
  >
    <div
      className="widget-google-maps-wrapper"
    >
      <div
        className="widget-google-maps"
      />
    </div>
  </div>
</LoadScript>
`;

exports[`Google maps renders a map with percentage of width and height units renders the structure correctly 1`] = `
<LoadScript
  googleMapsApiKey=""
  id="_com.mendix.widget.custom.Maps.Maps"
  onLoad={[Function]}
  version="weekly"
>
  <div
    className="widget-maps"
    style={
      Object {
        "height": "auto",
        "paddingBottom": "37.5%",
        "width": "50%",
      }
    }
  >
    <div
      className="widget-google-maps-wrapper"
    >
      <div
        className="widget-google-maps"
      />
    </div>
  </div>
</LoadScript>
`;

exports[`Google maps renders a map with pixels renders structure correctly 1`] = `
<LoadScript
  googleMapsApiKey=""
  id="_com.mendix.widget.custom.Maps.Maps"
  onLoad={[Function]}
  version="weekly"
>
  <div
    className="widget-maps"
    style={
      Object {
        "height": "75px",
        "width": "50px",
      }
    }
  >
    <div
      className="widget-google-maps-wrapper"
    >
      <div
        className="widget-google-maps"
      />
    </div>
  </div>
</LoadScript>
`;

exports[`Google maps renders a map with right structure 1`] = `
<LoadScript
  googleMapsApiKey=""
  id="_com.mendix.widget.custom.Maps.Maps"
  onLoad={[Function]}
  version="weekly"
>
  <div
    className="widget-maps"
    style={
      Object {
        "height": "37.5px",
        "width": "50px",
      }
    }
  >
    <div
      className="widget-google-maps-wrapper"
    >
      <div
        className="widget-google-maps"
      />
    </div>
  </div>
</LoadScript>
`;
